### ISSUE TITLE
No output with n_beams=1

### ISSUE THREAD

**Issue by melihyilmaz** [Noble-Lab/casanovo#185]
There are no PSMs recorded in the output `mztab` file when decoding is done with a single beam. The model seemed to be decoding garbage peptides (e.g. ending in the dummy token) that it ended up not caching at all. More details to come.

**Comment by wsnoble** [comment]
[one-beam.tgz](https://github.com/Noble-Lab/casanovo/files/11466725/one-beam.tgz)

Here is a zipped tar file containing a minimal example.  The `runall` script runs the decoding twice, once using one beam and one using five.  As you can see from the included mztab files, there is no output in the single-beam case.  Note that the example uses the most recent version of Casanovo (6299bd2), and it requires a ckpt file compatible with that. The script has a hard-coded pointer to such a file, which is too big to upload here.

**Comment by melihyilmaz** [comment]
I've identified the problem to be [this line](https://github.com/Noble-Lab/casanovo/blame/6299bd26b44d0d4c9b3c0bd20d23bcf595a4b0c2/casanovo/denovo/model.py#L612) which masks out all tokens when `n_beams=1` and behaves correctly otherwise. 

This is a strange bug but I think it has to do with `einops.repeat()` operating on the view of the tensor rather then a copy [here](https://github.com/Noble-Lab/casanovo/blob/6299bd26b44d0d4c9b3c0bd20d23bcf595a4b0c2/casanovo/denovo/model.py#L608).  Adding, `.clone()` to this line solves the issue but I wonder if @bittremieux any recommendations for a more efficient solution.

Otherwise, I'll implement that and set up a unit test for single beam decoding.

**Comment by bittremieux** [comment]
I don't think einops preferring views is the problem here. `repeat` generates a tensor that is bigger than the input (i.e. by repeating dimensions), so by definition this can't be a view of the original tensor. So I remain a bit confused about this issue.

Are you sure that `beams > 1` then actually works correctly? Maybe it returns `beams - 1` beams instead, but because that's more than zero, it still gives results and the bug is undiagnosed.

**Comment by melihyilmaz** [comment]
`beams > 1` works as intended with the correct number of beams returned.

You can find the simplest version of the weird behavior with tensor slicing after `repeat` in the below snippet and attached snapshot of output. 

- First cell shows correct behavior with `beams=2` , i.e. dummy token at zero indices are masked out. 
- Second cell shows buggy behavior with `beams=1`, i.e. all tokens are masked out instead of just the dummy token at the zero index.
- Third cell shows how `clone()` for repeated tensor get the correct behavior for `beams=1`.

```
import einops
import torch

beam=2
vocab=29
test_tensor = einops.repeat(torch.tensor([False,False]), "(B S) -> B (V S)", S=beam, V=vocab)
print(test_tensor)
test_tensor[:, :beam] = True
print(test_tensor)
```


![image](https://github.com/Noble-Lab/casanovo/assets/32707537/72d18457-e392-4b28-83d9-a261e1676eca)

**Comment by bittremieux** [comment]
Thanks for digging deeper and the MWE. After playing a bit with the code, I'm now convinced that you're completely right: the issue is indeed that `einops.repeat` returns a view, and using `clone` is the appropriate fix.
